common/models/templates/src/trie-model.ts

```diff
@@ -161,6 +161,116 @@
     public wordbreak(context: Context): USVString {
       return this.getLastWord(context.left);
     }
+
+    public getRootTraversal(): LexiconTraversal {
+      return new TrieModel.Traversal(this._trie['root'], '');
+    }
+
+    private static Traversal = class implements LexiconTraversal {
+      prefix: String;
+      root: Node;
+
+      constructor(root: Node, prefix: string) {
+        this.root = root;
+        this.prefix = prefix;
+      }
+
+      *children(): Generator<{char: string, traversal: () => LexiconTraversal}> {
+        let root = this.root;
+
+        if(root.type == 'internal') {
+          for(let i = 0; i < root.values.length; i++) {
+            let entry = root.values[i];
+            let entryNode = root.children[entry];
+
+            // SMP check
+            let charCode = entry.charCodeAt(0);
+            if(charCode >= 0xD800 && charCode <= 0xDBFF) {
+              // First part of a SMP char.
+              // For now, we'll just assume the second always completes such a char.
+              //
+              // Note:  Things get nasty here if this is only sometimes true; in the future,
+              // we should compile-time enforce that this assumption is always true if possible.
```

Comment:
TODO: always ensure every string in the models are USV strings (unicode scalar values); i.e., each surrogate pair is composed of exactly a high surrogate followed by a low surrogate.

Reply:
Is that a TODO for the lm-layer or for the model compiler?  Or both?

Reply:
This is a TODO for the model compiler, mostly!